vms/evm/uptimetracker/state.go

```diff
@@ -0,0 +1,234 @@
+	UpDuration    time.Duration `serialize:"true"`
+	LastUpdated   uint64        `serialize:"true"`
+	NodeID        ids.NodeID    `serialize:"true"`
+	Weight        uint64        `serialize:"true"`
+	StartTime     uint64        `serialize:"true"`
+	IsActive      bool          `serialize:"true"`
+	IsL1Validator bool          `serialize:"true"`
+
+	validationID ids.ID
+}
+
+// state holds the on-disk and cached representation of the validator set
+type state struct {
+	db database.Database
+
+	validators             map[ids.ID]*validator
+	nodeIDsToValidationIDs map[ids.NodeID]ids.ID
+	updatedValidators      set.Set[ids.ID]
+	deletedValidators      set.Set[ids.ID]
+}
+
+func newState(db database.Database) (*state, error) {
+	s := &state{
+		db:                     db,
+		validators:             make(map[ids.ID]*validator),
+		nodeIDsToValidationIDs: make(map[ids.NodeID]ids.ID),
+	}
+
+	it := db.NewIterator()
+	defer it.Release()
+
+	for it.Next() {
+		validationID, err := ids.ToID(it.Key())
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse validation ID: %w", err)
+		}
+
+		vdr := &validator{
+			validationID: validationID,
+		}
+
+		if _, err := codecManager.Unmarshal(it.Value(), vdr); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal validator: %w", err)
+		}
+
+		s.addValidator(validationID, vdr)
+	}
+
+	if err := it.Error(); err != nil {
+		return nil, fmt.Errorf("failed to iterate: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *state) GetUptime(nodeID ids.NodeID) (time.Duration, time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return 0, time.Time{}, database.ErrNotFound
+	}
+
+	return v.UpDuration, time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) SetUptime(
+	nodeID ids.NodeID,
+	upDuration time.Duration,
+	lastUpdated time.Time,
+) error {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return database.ErrNotFound
+	}
+
+	v.UpDuration = upDuration
+	v.LastUpdated = uint64(lastUpdated.Unix())
+
+	s.updatedValidators.Add(v.validationID)
+
+	return nil
+}
+
+func (s *state) GetStartTime(nodeID ids.NodeID) (time.Time, error) {
+	v, ok := s.getValidatorByNodeID(nodeID)
+	if !ok {
+		return time.Time{}, database.ErrNotFound
+	}
+
+	return time.Unix(int64(v.LastUpdated), 0), nil
+}
+
+func (s *state) addNewValidator(vdr *validator) {
+	s.addValidator(vdr.validationID, vdr)
+	s.updatedValidators.Add(vdr.validationID)
+}
+
+func (s *state) updateValidator(validationID ids.ID, isActive bool) bool {
+	v := s.validators[validationID]
+
+	if v.IsActive == isActive {
+		return false
+	}
+
+	v.IsActive = isActive
+	s.updatedValidators.Add(validationID)
+	return true
+}
+
+func (s *state) deleteValidator(validationID ids.ID) bool {
+	v, ok := s.validators[validationID]
+	if !ok {
+		return false
+	}
+
+	delete(s.validators, v.validationID)
+	delete(s.nodeIDsToValidationIDs, v.NodeID)
+
+	s.deletedValidators.Add(v.validationID)
+
+	return true
+}
+
+func (s *state) writeState() error {
```

Comment:
`state.writeState` seems weird. Should this be:
```suggestion
func (s *state) writeModifications() error {
```
or `flush` or something?